### PR TITLE
test(cli): add cli test cases, using bats

### PR DESCRIPTION
**Pull request by chanhsu001**
### What problem does this PR solve?

add skeleton code, add a few cases(ckb --version and ckb import,...) to verify design 

PR Summary: 
- Add ckb cli test cases, using bats platform.
- put bats test_cases into `cargo test`(cause these test cases need test-bed setup, so skip these cases when `cargo test` unless you use `-- --ignored` flag)
- add new CI workflow(`ci_ckb_cli`), process ckb cli bats test.
- test-bed setup including:
  - ckb binary build
  - bats and support lib installation
  - `[ckb-assets](https://github.com/nervosnetwork/ckb-assets)` for test db.

### Related changes

- PR to update `owner/repo`:

### Check List 

Tests 

- No code ci-runs-only: [ quick_checks,linters ]

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

---
This change is [Reviewable](https://reviewable.io/reviews/nervosnetwork/ckb/3336)

### CODE REVIEW

**Comment by chanhsu001** [comment]
```mermaid
graph LR
   b --> c
   c --> d
   subgraph Github Actions 
		subgraph ci_ckb_cli.yaml
			a[bats installation] --> b[ckb testbed env setup]
		end
	end
	
   c[devtools/ci_main.sh]

   subgraph cli.rs
	   d["#test<br> fn bats_suite{<br>...<br>}"]
	end
```

File: .github/workflows/ci_ckb_cli.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+      ) || (
+        github.event_name == 'workflow_dispatch'
+      )
+    steps:
+      - uses: actions/checkout@v2
+      - name: make and copy ckb binary to bats testbed directory
+        run: |
+          make build
+          mkdir -p /tmp/bats_testbed
+          cp target/release/ckb /tmp/bats_testbed/.
```
Comment (doitian) [inline .github/workflows/ci_ckb_cli.yaml:31]:
```suggestion
          cp target/release/ckb /tmp/bats_testbed/
```

File: util/app-config/src/tests/cli.rs [hunk 2]
```diff
@@ -97,3 +98,55 @@ fn ba_advanced() {
 
     assert_eq!(1, sub_matches.occurrences_of(ARG_BA_ADVANCED));
 }
+
+#[test]
+#[ignore]
+/// the test needs bats installation and pre test environment setup, so skip this test
+/// in normal case(eg. `cargo test`)
+fn bats_suite() -> Result<(), std::io::Error> {
```
Comment (doitian) [inline util/app-config/src/tests/cli.rs:106]: Why use rust to execute the test suite? Why not just use Makefile or a bash file?

File: .github/workflows/ci_ckb_cli.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      - 'pkg/*'
+  workflow_dispatch:
+
+env:
+  CARGO_TERM_COLOR: always
+  RUST_BACKTRACE: full
+  RUSTFLAGS: -D warnings
+jobs:
+  ckb_cli:
+    runs-on: ${{ self-hosted, linux }}
```
Comment (doitian) [inline .github/workflows/ci_ckb_cli.yaml:17]: `${{ }}` is a syntax to use a GitHub Action variable, here `self-hosted, linux` is not a valid variable name.

Please refer to `ci_integration_tests_ubuntu.yaml` how to set up the runner using a prologue script.

File: .github/workflows/ci_ckb_cli.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
+name: ci_ckb_cli
+on:
+  # Submit your review with a comment body containing "#ci_ckb_cli"
```
Comment (doitian) [inline .github/workflows/ci_ckb_cli.yaml:3]: Use the same condition as in `ci_integration_tests_ubuntu`.

**Review by doitian — DISMISSED** [review]
GitHub disallows running new CI job from a pull request. Please trigger the workflow in your clone and paste the passed job in the PR description.

**Comment by doitian** [comment]
bors d=liya2017

@liya2017 please help to review

File: util/app-config/src/tests/cli.rs (hunk 2 above)
Comment (chanhsu001) [reply util/app-config/src/tests/cli.rs:106]: most of tests in ckb are done by rust code(cargo test),  so i assume bats test follow this way.
yes, we could change into makefile/bash...

Comment (doitian) [reply util/app-config/src/tests/cli.rs:106]: It's unnecessary. Yes, just run bats from command line.

File: .github/workflows/ci_ckb_cli.yaml (hunk 4 above)
Comment (doitian) [reply .github/workflows/ci_ckb_cli.yaml:3]: This CI should be triggered just like integration tests.

File: .github/workflows/ci_ckb_cli.yaml (hunk 1 above)
Comment (chanhsu001) [reply .github/workflows/ci_ckb_cli.yaml]: done

File: util/app-config/src/tests/cli.rs (hunk 2 above)
Comment (chanhsu001) [reply util/app-config/src/tests/cli.rs]: now change mode to:  ci_yaml --> shell_do_real_job(cli_test.sh)--> call bats

File: .github/workflows/ci_ckb_cli.yaml (hunk 3 above)
Comment (chanhsu001) [reply .github/workflows/ci_ckb_cli.yaml]: done

File: .github/workflows/ci_ckb_cli.yaml (hunk 4 above)
Comment (chanhsu001) [reply .github/workflows/ci_ckb_cli.yaml]: add same prologue on ci_ckb_cli.yaml

File: .github/workflows/ci_ckb_cli.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+          repository: nervosnetwork/ckb-assets
+          ref: fd1030f93cdb06ae804b2af1255b903bcb2338f8
+
+      - name: init ckb bats testbed
+        run: |
+          mv cli_bats_env/ckb.json /tmp/bats_testbed/.
+          cd /tmp/bats_testbed && ./ckb init && ./ckb import ckb.json
+
+      - run: |
+          if [[ ${{ needs.prologue.outputs.os_skip }} == run ]] && [[ ${{ needs.prologue.outputs.job_skip }} == run ]];then
```
[the remark this answers is not in the document]
Comment (chanhsu001) [reply .github/workflows/ci_ckb_cli.yaml]: I don't actual sure about prologue works in deep, should I remove prologue now? @liya2017

**Comment by chanhsu001** [comment]
@liya2017 , please help to check it again...i pass the calling workflow under my repo (except the prologue part)
there still one problem:
how can I save the shell testing result, and upload to somewhere for later inspection?
it seems shell running result( console error message) cannot be shown on ci action log, Is there a better way to handle it?

**Comment by doitian** [comment]
bors try

**Comment by doitian** [comment]
It seems that the CI jobs are stuck in this PR. Trying via bors on https://github.com/nervosnetwork/ckb/commit/c2295a78e915c8fb3e0c15e6b5c3e04f89cb4a0b

**Comment by doitian** [comment]
> It seems that the CI jobs are stuck in this PR. Trying via bors on [c2295a7](https://github.com/nervosnetwork/ckb/commit/c2295a78e915c8fb3e0c15e6b5c3e04f89cb4a0b)

Passed on Ubuntu.

bors r=doitian,liya2017

**Comment by chanhsu001** [comment]
i correct the error in the ci_integration_ubuntu.yaml and  shfmt shell script in test....push again, let's see the result.

**Pull request merged**